nmrcraft/data/dataset.py

```diff
@@ -225,6 +237,80 @@ def preprocess_features(self, X):
         return X_scaled, scaler
 
+    def get_target_columns_separated(self):
+        """Returns the column indicies of the target array nicely sorted.
+        For example: metal_X1: [[0, 1], [1, 2, 3, 4]]"""
+        if (
+            "metal" in self.target_columns
+        ):  # If targets have metal, do weird stuff
+            metal_index = self.target_columns.index("metal")
+            y_column_indices = column_length_to_indices(
+                self.target_column_numbers
+            )
+            for i in range(len(y_column_indices)):
+                if i == metal_index:
+                    y_column_indices[i].append(y_column_indices[i][0] + 1)
+                if i > metal_index:
+                    y_column_indices[i] = [x + 1 for x in y_column_indices[i]]
+
+        elif "metal" not in self.target_columns:
+            y_column_indices = column_length_to_indices(
+                self.target_column_numbers
+            )
+        return y_column_indices
+
+    def more_than_one_target(self):
+        """Function returns true if more than one target is specified"""
+        return len(self.target_columns) > 1
+
+    def binarized_target_decoder(self, y):
+        """
+        function takes in the  target (y) array and transforms it back to decoded form.
+        For this function to be run the one-hot-preprocesser already has to have been run beforehand.
+        """
+        y_column_indices = column_length_to_indices(self.target_column_numbers)
+        ys = []
+        ys_decoded = []
+        # Split up compressed array into the categories
+        # If one-dimensional y (f.e only metals)
+        if isinstance(y[0], np.int64):
+            ys = list(y[:])  # copy list
+            ys = np.array(list(map(list, [[x] for x in ys])))
+            ys_decoded = self.encoders[0].inverse_transform(ys)
+            ys_decoded_properly_rotated = np.array(
+                list(map(list, [[x] for x in ys_decoded]))
+            )
+            # Decode the binarized metal using the original binarizer
+        # If multidimensional y
+        if not isinstance(y[0], np.int64):
+            for i in range(len(y_column_indices)):
+                ys.append(y[:, y_column_indices[i]])
+            # Decode the binarized categries using the original binarizers
+            for i in range(len(ys)):
+                ys_decoded.append(self.encoders[i].inverse_transform(ys[i]))
+            ys_decoded_properly_rotated = [
+                list(x) if i == 0 else x
+                for i, x in enumerate(map(list, zip(*ys_decoded)))
+            ]
+        return ys_decoded_properly_rotated
+
+    def confusion_matrix_data_adapter(self, y):
+        """
+        Takes in binary encoded target array and returns decoded flat list.
+        Especially designed to work with confusion matrix.
+        """
+        y_decoded = self.binarized_target_decoder(y)
+        flat_y_decoded = [y for ys in y_decoded for y in ys]
+        return flat_y_decoded
```

Comment:
That must have been technically difficult to implement! Thanks for taking care of this. In the final review, we could add a bit more detailed documentation here so that the TAs know what’s going on in this function.